File: backend/app/biosecurity_workflow/workflow_state_store.py
from __future__ import annotations

from dataclasses import dataclass, field
from threading import Lock


@dataclass
class TrackWorkflowState:
    workflow_id: str
    completed_steps: list[str] = field(default_factory=list)
    last_zone: str = ""
    updated_at: str = ""


class WorkflowStateStore:
    """In-memory workflow progress keyed by camera + track + workflow."""

    def __init__(self) -> None:
        self._states: dict[str, TrackWorkflowState] = {}
        self._lock = Lock()

    @staticmethod
    def _key(camera_id: str, track_id: int, workflow_id: str) -> str:
        return f"{camera_id}:{track_id}:{workflow_id}"

    def get_state(self, camera_id: str, track_id: int, workflow_id: str) -> TrackWorkflowState:
        key = self._key(camera_id, track_id, workflow_id)
        with self._lock:
            state = self._states.get(key)
            if state is None:
                state = TrackWorkflowState(workflow_id=workflow_id)
                self._states[key] = state
            return state

    def mark_step_completed(
        self,
        *,
        camera_id: str,
        track_id: int,
        workflow_id: str,
        step_code: str,
        zone_code: str,
        timestamp: str,
    ) -> TrackWorkflowState:
        state = self.get_state(camera_id, track_id, workflow_id)
        if step_code not in state.completed_steps:
            state.completed_steps.append(step_code)
        state.last_zone = zone_code
        state.updated_at = timestamp
        return state

    def reset(self, camera_id: str, track_id: int, workflow_id: str) -> None:
        key = self._key(camera_id, track_id, workflow_id)
        with self._lock:
            self._states.pop(key, None)


_store: WorkflowStateStore | None = None


def get_workflow_state_store() -> WorkflowStateStore:
    global _store
    if _store is None:
        _store = WorkflowStateStore()
    return _store

File: backend/src/modules/workflow/workflowStateStore.js
export class WorkflowStateStore {
  constructor() {
    /** @type {Map<string, { workflowId: string, completedSteps: string[], lastZone: string, updatedAt: string }>} */
    this._states = new Map()
  }

  /** @param {string} cameraId @param {number} trackId @param {string} workflowId */
  _key(cameraId, trackId, workflowId) {
    return `${cameraId}:${trackId}:${workflowId}`
  }

  /** @param {string} cameraId @param {number} trackId @param {string} workflowId */
  getState(cameraId, trackId, workflowId) {
    const key = this._key(cameraId, trackId, workflowId)
    let state = this._states.get(key)
    if (!state) {
      state = { workflowId, completedSteps: [], lastZone: '', updatedAt: '' }
      this._states.set(key, state)
    }
    return state
  }

  /**
   * @param {object} params
   * @param {string} params.cameraId
   * @param {number} params.trackId
   * @param {string} params.workflowId
   * @param {string} params.stepCode
   * @param {string} params.zoneCode
   * @param {string} params.timestamp
   */
  markStepCompleted({ cameraId, trackId, workflowId, stepCode, zoneCode, timestamp }) {
    const state = this.getState(cameraId, trackId, workflowId)
    if (!state.completedSteps.includes(stepCode)) {
      state.completedSteps.push(stepCode)
    }
    state.lastZone = zoneCode
    state.updatedAt = timestamp
    return state
  }

  /** @param {string} cameraId @param {number} trackId @param {string} workflowId */
  reset(cameraId, trackId, workflowId) {
    this._states.delete(this._key(cameraId, trackId, workflowId))
  }
}

let _store = null

export function getWorkflowStateStore() {
  if (!_store) {
    _store = new WorkflowStateStore()
  }
  return _store
}

export function resetWorkflowStateStore() {
  _store = new WorkflowStateStore()
}
